<h1>CIA_Factbook_SQL</h1>
<h3>Generate summary statistics using SQL</h3>
The Data is from CIA World Factbook (factbook.db) and was provided by DataQuest.

<h4>SQL: install and connect</h4>

In [10]:
!conda install -yc conda-forge ipython-sql

Solving environment: ...working... done

# All requested packages already installed.



In [6]:
%%capture
%load_ext sql
%sql sqlite:///factbook.db

<h4>Quick look at data</h4>

In [9]:
%%sql
SELECT *
  FROM facts
 LIMIT 5;

 * sqlite:///factbook.db
Done.


[(1, 'af', 'Afghanistan', 652230, 652230, 0, 32564342, 2.32, 38.57, 13.89, 1.51),
 (2, 'al', 'Albania', 28748, 27398, 1350, 3029278, 0.3, 12.92, 6.58, 3.3),
 (3, 'ag', 'Algeria', 2381741, 2381741, 0, 39542166, 1.84, 23.67, 4.31, 0.92),
 (4, 'an', 'Andorra', 468, 468, 0, 85580, 0.12, 8.13, 6.96, 0.0),
 (5, 'ao', 'Angola', 1246700, 1246700, 0, 19625353, 2.78, 38.78, 11.49, 0.46)]

- [name] - The name of the country.
- [area] - The total land and sea area of the country.
- [population] - The country's population.
- [population_growth] - The country's population growth as a percentage.
- [birth_rate] - The country's birth rate, or the number of births a year per 1,000 people.
- [death_rate] - The country's death rate, or the number of death a year per 1,000 people.
- [area] - The country's total area (both land and water).
- [area_land] - The country's land area in square kilometers.
- [area_water] - The country's waterarea in square kilometers.

<h4>Summary of data: min/max population, min/max population growth

In [12]:
%%sql
SELECT MIN(population) AS min_pop,
       MAX(population) AS max_pop,
       MIN(population_growth) AS min_pop_growth,
       MAX(population_growth) AS max_pop_growth
  FROM facts;

 * sqlite:///factbook.db
Done.


[(0, 7256490011, 0.0, 4.02)]

<h2>Population</h2>
<h4>Outliers: select countries with min population</h4>
(according to cell with summary there is a country without population)

In [111]:
%%sql
SELECT name AS Lowest_pop_country,
       population
  FROM facts
 WHERE population == (SELECT MIN(population)
                      FROM   facts);

 * sqlite:///factbook.db
Done.


[('Antarctica', 0)]

<h4>Outliers: select countries with max population</h4>
(according to cell above there is a country with 7 billion population)

In [110]:
%%sql
SELECT name AS Highest_pop_country,
       population
  FROM facts
 WHERE population == (SELECT MAX(population)
                      FROM   facts);

 * sqlite:///factbook.db
Done.


[('World', 7256490011)]

In [28]:
%%sql
SELECT name AS Highest_pop_country,
       population
  FROM facts
 WHERE population == (SELECT MAX(population)
                      FROM   facts
                      WHERE  name <> 'World');

 * sqlite:///factbook.db
Done.


[('China', 1367485388)]

<h2>Density</h2>
<h4>Get average population/country and average area/country</h4>
(excluding World)

In [108]:
%%sql
SELECT CAST(AVG(population) AS Integer) AS avg_pop,
       ROUND(AVG(area), 2) AS avg_area
  FROM facts
 WHERE name <> 'World';

 * sqlite:///factbook.db
Done.


[(32242666, 555093.55)]

<h4>Select high density countries</h4>
Countries with population above average, area below average

In [42]:
%%sql
  SELECT name,
         population,
         area
    FROM facts
   WHERE population > (SELECT AVG(population)
                       FROM   facts
                       WHERE  name <> 'World')
     AND area < (SELECT AVG(area)
                 FROM   facts
                 WHERE  name <> 'World')
ORDER BY population DESC
   LIMIT 10;

 * sqlite:///factbook.db
Done.


[('Bangladesh', 168957745, 148460),
 ('Japan', 126919659, 377915),
 ('Philippines', 100998376, 300000),
 ('Vietnam', 94348835, 331210),
 ('Germany', 80854408, 357022),
 ('Thailand', 67976405, 513120),
 ('United Kingdom', 64088222, 243610),
 ('Italy', 61855120, 301340),
 ('Korea, South', 49115196, 99720),
 ('Spain', 48146134, 505370)]

Top 5 countries with highest population density<br/>
(population_density = population / area)

In [45]:
%%sql
  SELECT name,
         population,
         area,
         population / area AS population_density
    FROM facts
ORDER BY population_density DESC
   LIMIT 10;

 * sqlite:///factbook.db
Done.


[('Macau', 592731, 28, 21168),
 ('Monaco', 30535, 2, 15267),
 ('Singapore', 5674472, 697, 8141),
 ('Hong Kong', 7141106, 1108, 6445),
 ('Gaza Strip', 1869055, 360, 5191),
 ('Gibraltar', 29258, 6, 4876),
 ('Bahrain', 1346613, 760, 1771),
 ('Maldives', 393253, 298, 1319),
 ('Malta', 413965, 316, 1310),
 ('Bermuda', 70196, 54, 1299)]

<h2>Growth Rate</h2>
<h4>Select countries with highest population and highest growth rate</h4>

In [33]:
%%sql
SELECT name AS Highest_growth_country,
       population_growth
  FROM facts
 WHERE population_growth == (SELECT MAX(population_growth)
                             FROM   facts
                             WHERE  name <> 'World');

 * sqlite:///factbook.db
Done.


[('South Sudan', 4.02)]

<h4>Calculate population for next year</h4>
new_population = population * population_growth

(show top 5 countries with highest expected population)

In [46]:
%%sql
  SELECT name,
         population,
         population_growth,
         CAST(population * population_growth AS Integer) AS new_population
    FROM facts
   WHERE name <> 'World'
ORDER BY new_population DESC
   LIMIT 5;

 * sqlite:///factbook.db
Done.


[('India', 1251695584, 1.22, 1527068612),
 ('China', 1367485388, 0.45, 615368424),
 ('Nigeria', 181562056, 2.45, 444827037),
 ('Pakistan', 199085847, 1.46, 290665336),
 ('Ethiopia', 99465819, 2.89, 287456216)]

<h2>Migration Rate</h2>
<h4>Outiers: highest migration rate</h4> 

In [79]:
%%sql
SELECT name AS highest_migration_rate_country,
       migration_rate
  FROM facts
 WHERE migration_rate == (SELECT MAX(migration_rate)
                          FROM   facts
                          WHERE  name <> 'World');

 * sqlite:///factbook.db
Done.


[('Qatar', 22.39)]

<h4>Count countries with migration_rate is equal to 0.0</h4>

In [100]:
%%sql
SELECT SUM(migration_rate=0.0) AS countries_zero_migration_rate
  FROM facts;

 * sqlite:///factbook.db
Done.


[(35,)]

<h4>Sort Countries according to migration_rate</h4>

In [106]:
%%sql
  SELECT name AS country,
         migration_rate,
         CASE
         WHEN migration_rate < (SELECT AVG(migration_rate) FROM facts) THEN 'low'
         WHEN migration_rate > (SELECT AVG(migration_rate) FROM facts) THEN 'high'
         END AS classified_migration_rate
    FROM facts
   WHERE migration_rate <> 'None'   
GROUP BY country
ORDER BY classified_migration_rate;

 * sqlite:///factbook.db
Done.


[('American Samoa', 21.13, 'high'),
 ('Anguilla', 12.18, 'high'),
 ('Armenia', 5.8, 'high'),
 ('Aruba', 8.92, 'high'),
 ('Australia', 5.65, 'high'),
 ('Austria', 5.56, 'high'),
 ('Bahrain', 13.09, 'high'),
 ('Belgium', 5.87, 'high'),
 ('Botswana', 4.56, 'high'),
 ('British Virgin Islands', 17.28, 'high'),
 ('Canada', 5.66, 'high'),
 ('Cayman Islands', 14.4, 'high'),
 ('Chad', 3.45, 'high'),
 ('Congo, Republic of the', 5.9, 'high'),
 ('Cuba', 3.66, 'high'),
 ('Cyprus', 9.48, 'high'),
 ('Djibouti', 6.06, 'high'),
 ('Dominica', 5.38, 'high'),
 ('El Salvador', 8.28, 'high'),
 ('Estonia', 3.6, 'high'),
 ('Fiji', 6.75, 'high'),
 ('Greenland', 5.98, 'high'),
 ('Guam', 6.34, 'high'),
 ('Guyana', 8.06, 'high'),
 ('Iceland', 4.43, 'high'),
 ('Ireland', 4.09, 'high'),
 ('Isle of Man', 6.56, 'high'),
 ('Italy', 4.1, 'high'),
 ('Jamaica', 4.66, 'high'),
 ('Jersey', 3.76, 'high'),
 ('Jordan', 13.24, 'high'),
 ('Kyrgyzstan', 5.22, 'high'),
 ('Latvia', 6.26, 'high'),
 ('Lesotho', 7.36, 'high'),
 ('Libya', 7.8, 'high'),
 ('Liechtenstein', 5.08, 'high'),
 ('Lithuania', 6.27, 'high'),
 ('Luxembourg', 17.16, 'high'),
 ('Maldives', 12.68, 'high'),
 ('Marshall Islands', 4.83, 'high'),
 ('Micronesia, Federated States of', 20.93, 'high'),
 ('Moldova', 9.67, 'high'),
 ('Monaco', 3.83, 'high'),
 ('Nauru', 13.63, 'high'),
 ('Nepal', 3.86, 'high'),
 ('New Caledonia', 4.01, 'high'),
 ('Northern Mariana Islands', 7.16, 'high'),
 ('Norway', 7.25, 'high'),
 ('Puerto Rico', 8.15, 'high'),
 ('Qatar', 22.39, 'high'),
 ('Saint Pierre and Miquelon', 8.49, 'high'),
 ('Saint Vincent and the Grenadines', 9.17, 'high'),
 ('Samoa', 9.78, 'high'),
 ('San Marino', 8.03, 'high'),
 ('Sao Tome and Principe', 8.63, 'high'),
 ('Singapore', 14.05, 'high'),
 ('Sint Maarten', 6.63, 'high'),
 ('Somalia', 8.49, 'high'),
 ('South Sudan', 11.47, 'high'),
 ('Spain', 8.31, 'high'),
 ('Sudan', 4.29, 'high'),
 ('Sweden', 5.42, 'high'),
 ('Switzerland', 4.74, 'high'),
 ('Syria', 19.79, 'high'),
 ('Timor-Leste', 3.86, 'high'),
 ('Tonga', 17.84, 'high'),
 ('Trinidad and Tobago', 6.25, 'high'),
 ('Turks and Caicos Islands', 9.94, 'high'),
 ('Tuvalu', 6.81, 'high'),
 ('United Arab Emirates', 12.36, 'high'),
 ('United States', 3.86, 'high'),
 ('Virgin Islands', 7.67, 'high'),
 ('Wallis and Futuna', 5.06, 'high'),
 ('Afghanistan', 1.51, 'low'),
 ('Albania', 3.3, 'low'),
 ('Algeria', 0.92, 'low'),
 ('Andorra', 0.0, 'low'),
 ('Angola', 0.46, 'low'),
 ('Antigua and Barbuda', 2.21, 'low'),
 ('Argentina', 0.0, 'low'),
 ('Azerbaijan', 0.0, 'low'),
 ('Bahamas, The', 0.0, 'low'),
 ('Bangladesh', 0.46, 'low'),
 ('Barbados', 0.3, 'low'),
 ('Belarus', 0.7, 'low'),
 ('Belize', 0.0, 'low'),
 ('Benin', 0.0, 'low'),
 ('Bermuda', 1.88, 'low'),
 ('Bhutan', 0.0, 'low'),
 ('Bolivia', 0.62, 'low'),
 ('Bosnia and Herzegovina', 0.38, 'low'),
 ('Brazil', 0.14, 'low'),
 ('Brunei', 2.43, 'low'),
 ('Bulgaria', 0.29, 'low'),
 ('Burkina Faso', 0.0, 'low'),
 ('Burma', 0.28, 'low'),
 ('Burundi', 0.0, 'low'),
 ('Cabo Verde', 0.63, 'low'),
 ('Cambodia', 0.32, 'low'),
 ('Cameroon', 0.15, 'low'),
 ('Central African Republic', 0.0, 'low'),
 ('Chile', 0.34, 'low'),
 ('China', 0.44, 'low'),
 ('Colombia', 0.64, 'low'),
 ('Comoros', 2.53, 'low'),
 ('Congo, Democratic Republic of the', 0.27, 'low'),
 ('Costa Rica', 0.83, 'low'),
 ("Cote d'Ivoire", 0.0, 'low'),
 ('Croatia', 1.39, 'low'),
 ('Curacao', 1.4, 'low'),
 ('Czech Republic', 2.33, 'low'),
 ('Denmark', 2.2, 'low'),
 ('Dominican Republic', 1.91, 'low'),
 ('Ecuador', 0.0, 'low'),
 ('Egypt', 0.19, 'low'),
 ('Equatorial Guinea', 0.0, 'low'),
 ('Eritrea', 0.0, 'low'),
 ('Ethiopia', 0.22, 'low'),
 ('European Union', 2.5, 'low'),
 ('Faroe Islands', 0.0, 'low'),
 ('Finland', 3.1, 'low'),
 ('France', 1.09, 'low'),
 ('French Polynesia', 0.84, 'low'),
 ('Gabon', 2.03, 'low'),
 ('Gambia, The', 2.12, 'low'),
 ('Gaza Strip', 0.0, 'low'),
 ('Georgia', 2.7, 'low'),
 ('Germany', 1.24, 'low'),
 ('Ghana', 2.02, 'low'),
 ('Gibraltar', 3.28, 'low'),
 ('Greece', 2.32, 'low'),
 ('Grenada', 3.13, 'l

<h2>Area</h2>
<h4>Select Countries that have more water than land</h4>
(ratio_water_land = area_water/area_land)

In [50]:
%%sql
SELECT name,
       area_water,
       area_land,
       (area_water / area_land) AS ratio_water_land
  FROM facts
 WHERE area_water > area_land
   AND name <> 'World';

 * sqlite:///factbook.db
Done.


[('British Indian Ocean Territory', 54340, 60, 905),
 ('Virgin Islands', 1564, 346, 4)]

British Indian Ocian Territory has 905 times more water than land<br/>
Virgin Islands has 4 times more water than land

<h4>Calculate percentage water</h4>
(perc_area_water = area_water / area)

In [132]:
%%sql
  SELECT name,
         area,
         area_land,
         area_water,
         ROUND((CAST(area_water AS Float) / CAST(area AS Float) * 100),2) AS perc_area_water
    FROM facts
   WHERE name <> 'World'
     AND perc_area_water <> 'None'
ORDER BY perc_area_water DESC;

 * sqlite:///factbook.db
Done.


[('British Indian Ocean Territory', 54400, 60, 54340, 99.89),
 ('Virgin Islands', 1910, 346, 1564, 81.88),
 ('Puerto Rico', 13791, 8870, 4921, 35.68),
 ('Bahamas, The', 13880, 10010, 3870, 27.88),
 ('Guinea-Bissau', 36125, 28120, 8005, 22.16),
 ('Malawi', 118484, 94080, 24404, 20.6),
 ('Netherlands', 41543, 33893, 7650, 18.41),
 ('Uganda', 241038, 197100, 43938, 18.23),
 ('Eritrea', 117600, 101000, 16600, 14.12),
 ('Liberia', 111369, 96320, 15049, 13.51),
 ('Bangladesh', 148460, 130170, 18290, 12.32),
 ('Gambia, The', 11300, 10120, 1180, 10.44),
 ('Taiwan', 35980, 32260, 3720, 10.34),
 ('Finland', 338145, 303815, 34330, 10.15),
 ('India', 3287263, 2973193, 314070, 9.55),
 ('Ethiopia', 1104300, None, 104300, 9.44),
 ('Canada', 9984670, 9093507, 891163, 8.93),
 ('Sweden', 450295, 410335, 39960, 8.87),
 ('Colombia', 1138910, 1038700, 100210, 8.8),
 ('Brunei', 5765, 5265, 500, 8.67),
 ('Guyana', 214969, 196849, 18120, 8.43),
 ('French Polynesia', 4167, 3827, 340, 8.16),
 ('Nicaragua', 130370, 119990, 10380, 7.96),
 ('Burundi', 27830, 25680, 2150, 7.73),
 ('Iran', 1648195, 1531595, 116600, 7.07),
 ('United States', 9826675, 9161966, 664709, 6.76),
 ('Tanzania', 947300, 885800, 61500, 6.49),
 ('Vietnam', 331210, 310070, 21140, 6.38),
 ('Rwanda', 26338, 24668, 1670, 6.34),
 ('Estonia', 45228, 42388, 2840, 6.28),
 ('Norway', 323802, 304282, 19520, 6.03),
 ('Armenia', 29743, 28203, 1540, 5.18),
 ('Tunisia', 163610, 155360, 8250, 5.04),
 ('Uzbekistan', 447400, 425400, 22000, 4.92),
 ('Indonesia', 1904569, 1811569, 93000, 4.88),
 ('Suriname', 163820, 156000, 7820, 4.77),
 ('Albania', 28748, 27398, 1350, 4.7),
 ('Ghana', 238533, 227533, 11000, 4.61),
 ('Azerbaijan', 86600, 82629, 3971, 4.59),
 ('Togo', 56785, 54385, 2400, 4.23),
 ('Russia', 17098242, 16377742, 720500, 4.21),
 ('Kyrgyzstan', 199951, 191801, 8150, 4.08),
 ('Tonga', 747, 717, 30, 4.02),
 ('Lithuania', 65300, 62680, 2620, 4.01),
 ('Ukraine', 603550, 579330, 24220, 4.01),
 ('West Bank', 5860, 5640, 220, 3.75),
 ('Gabon', 267667, 257667, 10000, 3.74),
 ('Turkmenistan', 488100, 469930, 18170, 3.72),
 ('Hungary', 93028, 89608, 3420, 3.68),
 ('Latvia', 64589, 62249, 2340, 3.62),
 ('Romania', 238391, 229891, 8500, 3.57),
 ('Japan', 377915, 364485, 13430, 3.55),
 ('Burma', 676578, 653508, 23070, 3.41),
 ('Congo, Democratic Republic of the', 2344858, 2267048, 77810, 3.32),
 ('Venezuela', 912050, 882050, 30000, 3.29),
 ('Pakistan', 796095, 770875, 25220, 3.17),
 ('Hong Kong', 1108, 1073, 35, 3.16),
 ('Solomon Islands', 28896, 27986, 910, 3.15),
 ('Switzerland', 41277, 39997, 1280, 3.1),
 ('Moldova', 33851, 32891, 960, 2.84),
 ('China', 9596960, 9326410, 270550, 2.82),
 ('Korea, South', 99720, 96920, 2800, 2.81),
 ('Poland', 312685, 304255, 8430, 2.7),
 ('Iceland', 103000, 100250, 2750, 2.67),
 ('Montenegro', 13812, 13452, 360, 2.61),
 ('Nepal', 147181, 143351, 3830, 2.6),
 ('Botswana', 581730, 566730, 15000, 2.58),
 ('Laos', 236800, 230800, 6000, 2.53),
 ('Cambodia', 181035, 176515, 4520, 2.5),
 ('Italy', 301340, 294140, 7200, 2.39),
 ('Ecuador', 283561, 276841, 6720, 2.37),
 ('Germany', 357022, 348672, 8350, 2.34),
 ('Paraguay', 406752, 397302, 9450, 2.32),
 ('Belarus', 207600, 202900, 4700, 2.26),
 ('Bulgaria', 110879, 108489, 2390, 2.16),
 ('Papua New Guinea', 462840, 452860, 9980, 2.16),
 ('Senegal', 196722, 192530, 4192, 2.13),
 ('Israel', 20770, 20330, 440, 2.12),
 ('Czech Republic', 78867, 77247, 1620, 2.05),
 ('Ireland', 70273, 68883, 1390, 1.98),
 ('Kenya', 580367, 569140, 11227, 1.93),
 ('Slovakia', 49035, 48105, 930, 1.9),
 ('Brazil', 8515770, 8358140, 157630, 1.85),
 ('Tajikistan', 144100, 141510, 2590, 1.8),
 ('Benin', 112622, 110622, 2000, 1.78),
 ('Turkey', 783562, 769632, 13930, 1.78),
 ('Austria', 83871, 82445, 1426, 1.7),
 ('Chile', 756102, 743812, 12290, 1.63),
 ('Lebanon', 10400, 10230, 170, 1.63),
 ('Mozambique', 799380, 786380, 13000, 1.63),
 ('Saint Lucia', 616, 606, 10, 1.62),
 ('Somalia', 637657, 627337, 10320, 1.62),
 ('New Caledonia', 18575, 18275, 300, 1.62